proposals/0621-linear-constraints.rst

```diff
@@ -0,0 +1,1108 @@
+through the function. Using linear constraints, however, such a
+function can be written as:
+
+::
+
+   read2AndDiscard ::  (Read n, Write n) %1 => MArray a n -> (Ur a, Ur a)
+   read2AndDiscard arr =
+        let !(Box x)  = read arr 0
+            !(Box y)  = read arr 1
+            !()       = free arr
+        in (x, y)
+
+The main way in which this differs from our previous function is that
+our array is no longer a linear resource - it is
+*unrestricted*. However, we maintain the guarantee that it is used in
+a way which does not violate linearity through the ``Read n`` and
+``Write n`` linear constraints. Here, ``n`` is a type-level tag used
+to identify the array. Accordingly, our type constructor for
+``MArray`` is parameterised by ``n``.
+
+The type signatures for  ``read``, ``free``, and ``Box`` are:
+
+::
+
+   read  :: Read n %1 => MArray a n -> Int -> Read n /\ Ur a
+
+   free :: (Read n, Write n) %1 => MArray a n -> ()
+
+   data c /\ a where
+     Box :: c %1 => a -> c /\ a
+
+i.e. ``read`` is a function which consumes a linear ``Read n``
+constraint, allowing us to read from the specified array index. It also
+returns a new ``Read n`` constraint, allowing us to subsequently read
+from the array again. Likewise, ``free`` consumes both a ``Read n``
+and a ``Write n`` constraint and introduces none, ensuring that we
+cannot read or write after freeing.
+
+Thus we eliminate the need to manually thread the ownership of the
+array through the function, whilst maintaining the guarantees of
+unique ownership via the linear constraints.
+
+For a more in-depth example along these lines, refer to section 4 of
+the paper_.
+
+The Linearly constraint
+^^^^^^^^^^^^^^^^^^^^^^^
+
+We can create a class, the paper_ calls it ``Linearly`` with the
+following API:
+
+::
+
+  %Dupable class Linearly
+
+  linearly :: (Linearly %1 => Ur a) %1 -> Ur a
+  newLinearlyDict :: Linear.IO (Dict Linearly)
+
+  data Dict c where
+    Dict :: c %1 => Dict c
+
+This ensures that it is not possible to ever build an unrestricted
+evidence for ``Linearly``.
+
+Equipped with this we can extend the API of the example above with
+a way to create arrays::
+
+  new :: Linearly %1 => Int -> NewMArray a
+
+  data NewMArray a where
+    NewMArray :: (Read n, Write n) %1 => MArray a n -> NewMArray a n
+
+Because there is no unrestricted evidence of ``Linearly``, the
+linearity of the ``Linearly`` constraint will contaminate the returned
+``NewMArray a`` value, ensuring in turn that the returned ``Read n``,
+and ``Write n`` constraints *must* be used linearly, as required.
+
+The difference with having ``new`` itself use a continuation is that
+we can now have several calls to ``new`` in the same scope. Which
+prevents the problems described in the *Motivation* section.
+
+::
+
+  linearly $
+    let
+      !(NewMArray arr1) = new
+      !(NewMArray arr2) = new
+    in
+    … -- modify the array as suited
+    Ur $ sum arr1 + sum arr2
```

Comment:
It would be good to explain somewhere (perhaps in the context of this example) why `Linearly` needs to be `Dupable`, as the paper does